pkg/trace/telemetry/collector.go

```diff
@@ -112,34 +118,37 @@ func NewNoopCollector() TelemetryCollector {
 	return &noopTelemetryCollector{}
 }
 
-func (f *telemetryCollector) sendEvent(event *OnboardingEvent) {
+func (f *telemetryCollector) sendEvent(event *OnboardingEvent) (err error) {
 	body, err := json.Marshal(event)
 	if err != nil {
-		return
+		return err
 	}
 	bodyLen := strconv.Itoa(len(body))
 	for _, endpoint := range f.endpoints {
-		req, err := http.NewRequest("POST", endpoint.Host, bytes.NewReader(body))
-		if err != nil {
+		req, reqErr := http.NewRequest("POST", endpoint.Host, bytes.NewReader(body))
+		if reqErr != nil {
+			err = reqErr
 			continue
 		}
 		req.Header.Add("Content-Type", "application/json")
 		req.Header.Add("User-Agent", f.userAgent)
 		req.Header.Add("DD-Api-Key", endpoint.APIKey)
 		req.Header.Add("Content-Length", bodyLen)
 
-		resp, err := f.client.Do(req)
-		if err != nil {
+		resp, reqErr := f.client.Do(req)
+		if reqErr != nil {
+			err = reqErr
 			continue
 		}
 		// Unconditionally read the body and ignore any errors
 		_, _ = io.Copy(io.Discard, resp.Body)
 		resp.Body.Close()
 	}
+	return err
```

Comment:
Looks like this will return an error even if some endpoint succeeded. Also it only returns the last error.

I see below we're checking if we've sent before. If there's a bad endpoint in the config, I think this will result in constant sending of first trace events.

Reply:
I just don't want us to drop the event if one of the endpoints has temporary issues

We could hope that this is not too common, or add limited retries

Reply:
Is it a problem that it's going to be submitted multiple times to the same endpoint?
If one endpoint fails, the successful endpoints will get continuous events until it stops failing. Not sure of your use case for the first trace event, but I would think it would make it more or less useless.

It's not that uncommon to have misconfigurations, or services being down. I think we should account for this.

At the very least, we should limit this in some way. We don't want a bad config or down service to trigger a telemetry submission on every trace write.

Reply:
All of our endpoints are idempotent on the backend: one and the same event is safe to send any number of times, the backend will correctly deduplicate them based on a hash of their contents.

You do have a good point that if one of the endpoints is down, the other one will get spammed with continual "first trace sent" errors. I've made a change to try sending the event no more than 5 times. If we fail, we bump a counter and once it reaches a certain count, we stop retrying.

Reply:
> All of our endpoints are idempotent on the backend: one and the same event is safe to send any number of times, the backend will correctly deduplicate them based on a hash of their contents.

Ok, good to know. It looks like time isn't a parameter, so it will be the same payload every time, even if it's triggered later or by a different payload?

> You do have a good point that if one of the endpoints is down, the other one will get spammed with continual "first trace sent" errors. I've made a change to try sending the event no more than 5 times. If we fail, we bump a counter and once it reaches a certain count, we stop retrying.

This looks fine.

Reply:
Correct, this is just the time the agent was originally installed, not the time the trace is being sent.